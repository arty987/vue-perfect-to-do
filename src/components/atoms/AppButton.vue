<script>
export default {
  name: "AppButton"
}
</script>

<template>
  <button type="button">
    <slot></slot>
  </button>
</template>