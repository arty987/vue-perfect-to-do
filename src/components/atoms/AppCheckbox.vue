<script>
export default {
  name: "AppCheckbox",
  props: {
    disabled: {
      type: Boolean,
      default: false
    },
    checked: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<template>
  <input
      type="checkbox"
      :disabled="disabled"
      :checked="checked"
  />
</template>