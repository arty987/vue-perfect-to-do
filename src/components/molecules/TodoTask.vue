<script>
import {mapActions} from "vuex";
import AppCheckbox from "@/components/atoms/AppCheckbox.vue";
import AppButton from "@/components/atoms/AppButton.vue";

export default {
  name: "TodoTask",
  components: {
    AppButton,
    AppCheckbox
  },
  props: {
    task: {
      type: Object,
      required: true,
    }
  },
  methods: {
    ...mapActions([
      "removeTask",
      "reverseCompleted",
    ])
  }
}
</script>

<template>
  <li>
    <AppCheckbox :checked="task.completed" @change="reverseCompleted(task)"/>
    <div>{{ task.title }}</div>
    <AppButton @click="removeTask(task)">X</AppButton>
  </li>
</template>

<style scoped>

</style>