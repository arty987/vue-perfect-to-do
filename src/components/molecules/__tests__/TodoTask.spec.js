import {shallowMount} from '@vue/test-utils'
import TodoTask from '@/components/molecules/TodoTask.vue';
import AppCheckbox from '@/components/atoms/AppCheckbox.vue';
import {createStore} from 'vuex';
import AppButton from '@/components/atoms/AppButton.vue';

describe('TodoTask.vue', () => {

    const task = {
        id: 1,
        title: 'Task 1',
        completed: true
    };

    let actions;
    let store;

    beforeEach(() => {
        actions = {
            removeTask: jest.fn(),
            reverseCompleted: jest.fn()
        };

        store = createStore({
            actions
        });
    })

    it.each([
        true, false
    ])('Should render TodoTask', (completed) => {
        const wrapper = shallowMount(TodoTask, {
            props: {
                task: {
                    ...task,
                    completed
                }
            }
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it('Should dispatch reverseCompleted action on AppCheckbox change', async () => {
        const wrapper = shallowMount(TodoTask, {
            props: {
                task
            },
            global: {
                plugins: [store]
            }
        });

        await wrapper.findComponent(AppCheckbox).trigger('change');

        expect(actions.reverseCompleted).toHaveBeenCalledWith(expect.any(Object), task);
    });

    it('Should dispatch removeTask action on AppButton click', async () => {
        const wrapper = shallowMount(TodoTask, {
            props: {
                task
            },
            global: {
                plugins: [store]
            }
        });

        await wrapper.findComponent(AppButton).trigger('click');

        expect(actions.removeTask).toHaveBeenCalledWith(expect.any(Object), task);
    });
})