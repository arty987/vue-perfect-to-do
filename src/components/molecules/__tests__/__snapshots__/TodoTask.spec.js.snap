// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoTask.vue Should render TodoTask 1`] = `
<li
  class="list-group-item d-flex justify-content-between align-items-center"
>
  <div
    class="d-flex flex-row align-items-center"
  >
    <app-checkbox-stub
      checked="true"
      disabled="false"
    />
    <div
      class="task-title completed"
    >
      Task 1
    </div>
  </div>
  <app-button-stub
    class="btn-close"
  />
</li>
`;

exports[`TodoTask.vue Should render TodoTask 2`] = `
<li
  class="list-group-item d-flex justify-content-between align-items-center"
>
  <div
    class="d-flex flex-row align-items-center"
  >
    <app-checkbox-stub
      checked="false"
      disabled="false"
    />
    <div
      class="task-title"
    >
      Task 1
    </div>
  </div>
  <app-button-stub
    class="btn-close"
  />
</li>
`;
